internal/cli/status.go

```diff
@@ -146,6 +156,458 @@ func (c *StatusCommand) Run(args []string) int {
 	return 0
 }
 
+// RefreshApplicationStatus takes a project and application target and generates
+// a list of applications to refresh the status on. If all projects are requested
+// to be refreshed, the CLI will do its best to honor the request. However if
+// a project is local and the CLI was not invoked inside that project dir, the
+// CLI won't be able to refresh that projects application statuses.
+func (c *StatusCommand) RefreshApplicationStatus(projectTarget, appTarget string) error {
+	// Get our API client
+	client := c.project.Client()
+
+	// Get the entire list of apps
+	// Determine project locality
+	// Do the Work (local or remote)
+	if c.flagAllProjects {
+		resp, err := client.ListProjects(c.Ctx, &empty.Empty{})
+		if err != nil {
+			c.ui.Output("Failed to retrieve all projects:"+clierrors.Humanize(err), terminal.WithErrorStyle())
+			return err
+		}
+		allProjectsRef := resp.Projects
+
+		for _, pRef := range allProjectsRef {
+			projectResp, err := client.GetProject(c.Ctx, &pb.GetProjectRequest{
+				Project: pRef,
+			})
+			if err != nil {
+				return err
+			}
+			project := projectResp.Project
+
+			workspace, err := c.getWorkspaceFromProject(projectResp)
+			if err != nil {
+				return err
+			}
+
+			// local or remote project?
+			var remote bool
+			if project.DataSource != nil {
+				switch project.DataSource.Source.(type) {
+				case *pb.Job_DataSource_Local:
+					remote = false
+				case *pb.Job_DataSource_Git:
+					remote = true
+				default:
+					remote = false
+				}
+			}
+
+			if !remote {
+				c.ui.Output("Cannot refresh every projects status without command being "+
+					"invoked inside project directory due to project being a local data source. "+
+					"Will attempt to refresh all project statuses anyway, but expect errors.",
+					terminal.WithWarningStyle())
+			}
+
+			if err = c.doRefresh(project, project.Applications, workspace, remote); err != nil {
+				c.ui.Output("Failed to refresh all statuses on project %q:"+clierrors.Humanize(err),
+					project.Name, terminal.WithErrorStyle())
+				return err
+			}
+		}
+	} else {
+		projectResp, err := client.GetProject(c.Ctx, &pb.GetProjectRequest{
+			Project: &pb.Ref_Project{
+				Project: projectTarget,
+			},
+		})
+		if err != nil {
+			if status.Code(err) == codes.NotFound {
+				var serverAddress string
+				if c.serverCtx != nil {
+					serverAddress = c.serverCtx.Server.Address
+				}
+
+				c.ui.Output(wpProjectNotFound, projectTarget, serverAddress, terminal.WithErrorStyle())
+			}
+			return err
+		}
+		project := projectResp.Project
+
+		// local or remote project?
```

Comment:
I don't think we should be making this distinction here. We should just schedule a status report job and let the CLI handle this (in the same way `waypoint build` does). The command itself shouldn't worry about this.

Reply:
The automatic runner allocation must be achieved through the `DoApp` function? If I remember right I had trouble making that work when the CLI is configured to have an optional config like this one, but I can give it a shot! If that works then this just turns into a loop calling DoApp on each app and avoiding the `ExpediteStatusReport` endpoint all together.

Reply:
I don’t remember the details anymore but also happy to help change how any of that works to make it work for you. I think that logic just makes more sense at the core CLI layer rather than the subcommand business logic.

Reply:
Agreed...I think that'll make this implementation a lot shorter and cleaner too. And in the future when we fix remote runners being able to refresh statuses this command doesn't change. I'll give it a go and see what I was missing last time and if I can fix it! 👍🏻

Reply:
Alright, got this updated to do `DoApp` in both instances!